src/dmd/glue.d

```diff
@@ -1163,8 +1163,72 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
 
     if (fd.isCrtCtorDtor & 1)
         objmod.setModuleCtorDtor(s, true);
+
     if (fd.isCrtCtorDtor & 2)
-        objmod.setModuleCtorDtor(s, false);
+    {
+        /*
+            https://issues.dlang.org/show_bug.cgi?id=22520
+
+            Apple radar: https://openradar.appspot.com/FB9733712
+
+            Apple deprecated the mechanism used to implement `crt_constructor`
+            on MacOS Monterey. This works around that by generating a new function
+            (crt_destructor_thunk_NNN, run as a constructor) which registers
+            the destructor-to-be using __cxa_atexit()
+
+            This workaround may need a further look at when it comes to
+            shared library support, however there is no bridge for
+            that spilt milk to flow under yet.
+
+            This relies on the Itanium ABI so is portable to any
+            platform it, if needed.
+        */
+        if(!target.c.crtDestructorsSupported)
+        {
+            __gshared uint nthDestructor = 0;
+            char* buf = cast(char*) calloc(50, 1);
+            assert(buf);
+            const ret = snprintf(buf, 100, "_dmd_crt_destructor_thunk.%u", nthDestructor++);
+            assert(ret >= 0 && ret < 100, "snprintf either failed or overran buffer");
+            //Function symbol
+            auto newConstructor = symbol_calloc(buf);
+            //Build type
+            newConstructor.Stype = type_function(TYnfunc, [], false, type_alloc(TYvoid));
+            //Tell it it's supposed to be a C function. Does it do anything? Not sure.
+            newConstructor.Stype.Tmangle = mTYman_c;
```

Comment:
I think you probably need to set this for `__dso_handle` too.

Reply:
I'm not convinced this actually does anything. If I let it generate the symbol itself on Linux it links just fine. I was reading an LLD thread and I think the symbol is supposed to be generated by the linker on MacOS

Reply:
> I'm not convinced this actually does anything. If I let it generate the symbol itself on Linux it links just fine. I was reading an LLD thread and I think the symbol is supposed to be generated by the linker on MacOS

Yes it does, for instance ensures that the proper mangling is done on the symbol.